### ISSUE TITLE
Trusted per Department

### ISSUE THREAD

**Issue by bds002** [magfest/ubersystem#1474]
Related to #1473 If we are looking to make some changes to the interactions of the Staffing section of code in regards to Staffing and Registration, it would be a good time to look at making staffers trusted per department.

The trusted tag means different things to different department heads:
--Game room used the trusted tag to differentiate who can be in charge, or take on extra responsibility within their department.  
--Treasury uses it to distinguish who can handle cash on site.
-- Merch uses as a limitation as to who can do what within their department.
-- Registration uses it much the same way.

We have a number of staffers that like to work across multiple departments, and just because someone is trusted in the gameroom, doesn't mean they should be picking up restricted shifts in Merch or at registration.

I don't see an easy fix for this but it has cause a couple of issues at the last 2 MAGFest Prime/Super events.

**Comment by binary1230** [comment]
I have created a user story for this at 
https://jira.magfest.net/browse/INC-128

I have marked this as high priority in JIRA and we'll try to get to it before this event, a lot of people have been asking for it

**Comment by thaeli** [comment]
Can we please keep development discussion in GitHub? This is a prime example of an issue we would move to upstream, I think.

**Comment by binary1230** [comment]
oh hey yea definitely, I can see how you might have thought otherwise from my comments. let me clarify:

I agree and think anything development-related, for the ease of everyone involved magfest or not, should stay on github.

If it's magfest-specific like config or server-specific or whatever, that's stuff we'll move to JIRA.

In the case of some issues like this one, the tech discussion will stay on github like always but might be mirrored by a JIRA issue which references the github issue for our internal planning.

the goal is to prune the github issues of anything magfest-specific like "need Gameroom dept renamed to Consoles"

**Comment by binary1230** [comment]
and yea my goal is that, when we have pruned this of magfest-specific issues which get moved to JIRA, this entire repo (and all of its issues) could just become the upstream after the repo is renamed to rams/core or similar.

**Comment by thaeli** [comment]
Cool, that sounds good. You are correct that I got a different impression from the previous comment.. Thanks for the clarification.

**Comment by binary1230** [comment]
To keep things simple, I'm going to kill the JIRA-related discussion here since @thaeli and I are now on the same page.

**Comment by binary1230** [comment]
Going to start working on this since it's been identified as a big issue for a lot of different departments.

The question is, what specifically do we do here? Thoughts in the next comment:

**Comment by binary1230** [comment]
My first initial guess at a solution for this would be to change this checkbox on the edit staffer page from this:

![image](https://cloud.githubusercontent.com/assets/5413064/10542264/012cb03c-73e5-11e5-84a3-d562145fea77.png)


to something like this:

![image](https://cloud.githubusercontent.com/assets/5413064/10542297/338a8a4a-73e5-11e5-8171-b0656e6260fe.png)

the [extra text] section would say:

"Check which departments can this staffer sign up for restricted shifts"

----------

On the backend, we would have to figure out how to map people with existing 'trusted' statuses to per-department trusted status. (we'll also need to write a DB migration for this)  

We might naively do this by just taking anyone with a trusted status and making them be trusted in whichever departments they are currently associated with, which is the behavior that we have right now.  Then, STOPs and departments can take a pass on their trusted staff and cull out anyone they don't currently recognize.

-------

Am I missing anything? glossing over anything?

**Comment by binary1230** [comment]
btw, that list for the trusted depts would be the same exact items as the 'assigned departments' list.

**Comment by EliAndrewC** [comment]
Ehhh, what you suggested is probably the best way to get something working that we'll actually have time for this year.  The REAL way that I'd like to handle this is to have department assignments no longer be a ``MultiChoice`` field and instead be a many-to-many relationship with tables something like this:

```python
class Department(MagModel):
    name = Column(UnicodeText)
    slug = Column(UnicodeText)
    parent = Column(UUID, ForeignKey('department.id'), nullable=True)
    shift_types = Column(MultiChoice(c.SHIFT_TYPE_OPTS))

class JobAssignment(MagModel):
    dept_id = Column(UUID, ForeignKey('department.id'))
    attendee_id = Column(UUID, ForeignKey('attendee.id'))
    roles = Column(MultiChoice(c.ROLE_OPTS))
    allowed_shifts = 
```

This general approach allows us the flexibility to do a bunch of stuff we've been wanted to do for awhile:
- make "department head" a per-department thing, since some people are assigned to multiple departments but are only the head of one, or some people are the heads of multiple, etc
- split out "department head" into more fine-grained definitions; some people are the point of contact for a department w.r.t. emails, some people are managers on-site, some people make policy decisions, etc. (this is a real problem for the department head checklist, where STOPS is waiting on people to fill it out, but those people don't think it's their job to do so)
- makes trusted a per-department thing (the main point of this particular issue)
- splits "trusted" into a non-binary definition, since just "is this shift restricted or not" is too course to be useful for some departments, who literally have to make sub-departments to work around those limitations
- allows us to define some departments as being child-departments, e.g. there are contractors who get their own department who it would be nice to treat as sub-departments of Consoles, etc
- tracks departments in the database, allowing us to create new departments and edit department names in the UI

I'd hoped to tackle this for Mag2016, but with all the other stuff I took on, I just haven't had the time, and don't see this as a reasonable goal until after Mag2016.

So the question is whether we want to try implementing something like what Dom suggested in the meantime.  I'm personally of the opinion that we can hobble along with a global trusted for this year and then fix it next year, since it would be inefficient to spend time on implementing something we're going to throw away in the medium term.  With that being said, if we DO decide to tackle this issue this year, I think Dom's suggestion is perfect in terms of how we'd go about it, so that's my 2 cents.

**Comment by bds002** [comment]
While Eli and I are on the same page 99% of the time with regards to Staffing issues.... on this one I disagree.  I think this has become a big enough issues for department heads that if we have the ability to do a intermediate step, then we should do that.  The conversation out of the Staffing Operations office to Dh's about this issue and other issues continues to be "we are working on it" with no real change in work flow. 

There are some other changes I would added to Eli's work flow above for a massive update to this process. Mainly separating out DH's ability to add registration entries or edit a staffers registration info. (Make it view able in a read only state).

**Comment by EliAndrewC** [comment]
I'll defer to Brent on how urgent this is; if he thinks that it's worth the time and effort to implement the intermediate step for this year then I say go for it.

**Comment by EliAndrewC** [comment]
(Also, I'll definitely consult with people, especially with Brent, before actually implementing any of what I described above, since that was a rough-draft and not my idea of the final project, e.g. the idea about adding access controls is a good one.)

**Comment by binary1230** [comment]
I figured this would probably be a big discussion :)

Sounds like we have uncovered a whole bunch of larger user stories here, would love to sit down sometime and brainstorm all those out so we have this task served up nicer for a downtime period like after mag16.

I -think- given that the large-scale task is pretty major, and the intermediate might turn out to be pretty simple, and depts have been asking for this for a while, the intermediate step might be a good option.  It buys us a lot of value while also being easy to rip out later.
